### PR TITLE
Add more... button

### LINKED ISSUE #87: Secondary buttons should collapse into a 'more...' button when there is not enough horizontal space.

### PR DESCRIPTION
**Pull request by gkchestertron**
### This project uses [semver](semver.org), please check the scope of this pr:

- [ ] #none# - documentation fixes and/or test additions
- [ ] #patch# - backwards-compatible bug fix
- [ ] #minor# - adding functionality in a backwards-compatible manner
- [x] #major# - incompatible API change

Closes #87 

# CHANGELOG

* Added "More..." button
  * Added default behavior that collapses extra buttons when more than 4
  * Added ability to override default
  * Added docs for "More..." button
  * Added ability to change text of "More..." button
* Added `alwaysVisible` flag to `frost-action-bar`

### CODE REVIEW

**Comment by gkchestertron** [comment]
image

**Comment by gkchestertron** [comment]
20pjmrgr8s

File: addon/components/frost-action-bar.js [hunk 1]
```diff
@@ -67,38 +86,131 @@ export default Component.extend({
...
-  @computed('controls')
-  isControlsObject (controls) {
-    return typeOf(controls) === 'object'
+  @computed('controls', 'selectedItems.[]', 'moreActionButtons')
+  /**
+   * gets controls for the action bar
+   * @param {array|object} controls - array or hash of controls
+   * @param {array} selectedItems - array of selected items for filtering controls
+   * @param {boolean|object|array} moreActionButtons - array or hash of extra controls
+   *  or flag indicating whether to generate them automaticaly
```
Comment (lirwin) [inline addon/components/frost-action-bar.js]: change to `automatically`

File: README.md [hunk 2]
```diff
@@ -26,13 +26,16 @@ ember install ember-frost-object-browser
...
-| loadingText      | `string`      |       | Text that appears beside the loading animation      |
-| selectedItems    | `array`       |       | List of items that are currently selected           |
+| Attribute             | Type          | Value | Description                                                 |
+| --------------------- | ------------- | ----- | ----------------------------------------------------------- |
+| controls              | `array`       |       | Controls that will be available in the action bar           |
+| isLoading             | `boolean`     | false | **default** - Action bar is not in a loading state          |
+|                       |               | true  | Action bar is in a loading state                            |
+| isVisible             | `boolean`     |       | defaults to computed prop based on selectedItems emptiness  |
+| loadingText           | `string`      |       | Text that appears beside the loading animation              |
+| moreActionButtons     | `array`       |       | Controls that will be wrapped up into 'More...' button      |
```
Comment (lirwin) [inline README.md]: moreActionButtons  type is `array|boolean|object`

File: addon/components/frost-action-bar.js [hunk 3]
```diff
@@ -31,6 +32,12 @@ export default Component.extend({
     i18n: PropTypes.shape({
       formatSelectedItemsLabel: PropTypes.func.isRequired
     }),
+    moreActionButtons: PropTypes.oneOfType([
+      PropTypes.arrayOf(PropTypes.object),
```
Comment (lirwin) [inline addon/components/frost-action-bar.js]: PropTypes.arrayOf(PropTypes.EmberComponent)

File: README.md [hunk 4]
```diff
@@ -26,13 +26,16 @@ ember install ember-frost-object-browser
...
-| Attribute        | Type          | Value | Description                                         |
-| ---------------- | ------------- | ----- | --------------------------------------------------- |
-| controls         | `array`       |       | Controls that will be available in the action bar   |
-| isLoading        | `boolean`     | false | **default** - Action bar is not in a loading state  |
-|                  |               | true  | Action bar is in a loading state                    |
-| loadingText      | `string`      |       | Text that appears beside the loading animation      |
-| selectedItems    | `array`       |       | List of items that are currently selected           |
+| Attribute             | Type          | Value | Description                                                 |
+| --------------------- | ------------- | ----- | ----------------------------------------------------------- |
+| controls              | `array`       |       | Controls that will be available in the action bar           |
```
Comment (lirwin) [inline README.md]: `controls` type is `EmberComponent[]|object`

File: README.md [hunk 5]
```diff
@@ -26,13 +26,16 @@ ember install ember-frost-object-browser
...
+| isLoading             | `boolean`     | false | **default** - Action bar is not in a loading state          |
+|                       |               | true  | Action bar is in a loading state                            |
+| isVisible             | `boolean`     |       | defaults to computed prop based on selectedItems emptiness  |
+| loadingText           | `string`      |       | Text that appears beside the loading animation              |
+| moreActionButtons     | `array`       |       | Controls that will be wrapped up into 'More...' button      |
+| moreActionButtonsText | `string`      |       | Text for 'More...' button                                   |
+| selectedItems         | `array`       |       | List of items that are currently selected                   |
```
Comment (lirwin) [inline README.md]: do we not give info on array of what?  i think it is more useful to say
`EmberObject[]|object[]`  or whatever the correct syntax is rather than just `array`

File: addon/components/frost-action-bar.js [hunk 6]
```diff
@@ -67,38 +86,131 @@ export default Component.extend({
...
   },
 
   @readOnly
-  @computed('controls')
-  isControlsObject (controls) {
-    return typeOf(controls) === 'object'
+  @computed('controls', 'selectedItems.[]', 'moreActionButtons')
+  /**
+   * gets controls for the action bar
+   * @param {array|object} controls - array or hash of controls
```
Comment (lirwin) [inline addon/components/frost-action-bar.js]: same here
`{EmberComponent[]|object}` isn't it preferred to say array of what?

File: addon/components/frost-action-bar.js [hunk 7]
```diff
@@ -67,38 +86,131 @@ export default Component.extend({
...
+   * @param {array} selectedItems - array of selected items for filtering controls
```
Comment (lirwin) [inline addon/components/frost-action-bar.js]: `{EmberObject[]|object[]} selectedItems`  won't keep repeating...

File: addon/components/frost-action-bar.js [hunk 8]
```diff
@@ -67,38 +86,131 @@ export default Component.extend({
...
 
     return controls
   },
 
-  // == Functions =============================================================
-
+  /**
+   * filters out inapplicable controls
+   * @param {array} selectedItems - currently selected items
+   * @returns {array} - array of applicable controls
```
Comment (lirwin) [inline addon/components/frost-action-bar.js]: `* @returns {object} - object of applicable controls by itemType??`
this is returning an object, whose data shape it would be nice to see an example of...

add guard against `selectedItems` undefined
ln214
```javascript
selectedTypesWithControls (selectedItems) {
   if (isEmpty(selectedItems)) {
      return {}
   }
```

also for lines 222 - 226 add guard against `componentKeyNames` undefined
```javascript
if (isPresent(itemType) && isPresent(componentKeyNames)) {
          const itemTypeContent = getWithDefault(componentKeyNamesForTypes, itemType, {})
          const itemTypeContentControls = getWithDefault(itemTypeContent, get(componentKeyNames, 'controls'), [])
          typesWithControls[itemType] = itemTypeContentControls
        }
```

File: tests/integration/components/frost-action-bar-test.js [hunk 9]
```diff
@@ -203,4 +207,154 @@ describe(test.label, function () {
...
+        `)
+      })
+
+      it('should put moreActionButtons into moreActionButtons button', function () {
+        const $moreButton = $hook('moreActions')
+        expect($moreButton.find('li').length).to.equal(4)
+      })
+
+      it('should not put any controls into the moreActionButtons', function () {
+        // 6 becuase 5 buttons plus 1 more button
```
Comment (lirwin) [inline tests/integration/components/frost-action-bar-test.js]: typo s.b. `because`

File: README.md (hunk 2 above)
Comment (lirwin) [inline README.md]: struggling with `moreActionButtons`  as it is a `button` with `more actions`

would it be more appropriately named `moreActionsButton`?  and `moreActionsButtonText`?

File: addon/components/frost-action-bar.js (hunk 7 above)
Comment (notmessenger) [reply addon/components/frost-action-bar.js]: `{Ember.Object[]|object[]} selectedItems`

File: README.md (hunk 4 above)
Comment (notmessenger) [reply README.md]: `Ember.Component[]|object`

File: README.md (hunk 2 above)
Comment (notmessenger) [reply README.md]: `Ember.Component[]|boolean|object`

File: addon/components/frost-action-bar.js (hunk 3 above)
Comment (notmessenger) [reply addon/components/frost-action-bar.js]: This would be in place of, not in addition to.  The _README.md_ changes say
> The `moreActionButtons` property only accepts `frost-buttons`.

File: addon/components/frost-action-bar.js [hunk 10]
```diff
@@ -57,7 +65,18 @@ export default Component.extend({
           return `${count} ${items} selected`
         }
       },
-      isLoading: false
+      isLoading: false,
+
+      @readOnly
+      @computed('selectedItems.[]')
```
Comment (notmessenger) [inline addon/components/frost-action-bar.js]: Several things about this:

1) There is already an `isVisible` property on Ember Components so you should not name yours the same - https://www.emberjs.com/api/ember/2.18/classes/Component/properties/isVisible?anchor=isVisible

2) Because this computed property is `@readOnly` you will not be able to override it by passing it in via the `isVisible` property in the component template usage as believed.

3)  I am pretty certain that `ember-prop-types` will not set this computed property up as desired since it is included as part of `getDefaultProps()`

4) This approach is not a Ciena-supported one.  We do not want to support the ability to override a computed property definition, which is why we use `@readOnly`.  What should be done instead is support a property that can be used to toggle visibility.  Then you will still have a computed property that will take the existence and value of that property into account within its own definition, as well as the `selectedItems.[]` content.

File: addon/templates/components/frost-action-bar.hbs [hunk 11]
```diff
@@ -11,7 +11,44 @@
     <div class='loading-text'>{{loadingText}}</div>
   {{/if}}
 </div>
-<div class='{{css}}-buttons'>
+<div class='{{css}}-buttons' data-test={{hook 'frost-action-bar-buttons'}}>
```
Comment (notmessenger) [inline addon/templates/components/frost-action-bar.hbs]: Believe this will throw an console warning (error?) about needing to use html string safe or similar?

File: addon/components/frost-action-bar.js (hunk 10 above)
Comment (notmessenger) [reply addon/components/frost-action-bar.js]: It seems there's no test that is testing the scenario you desired?

File: addon/components/frost-action-bar.js [hunk 12]
```diff
@@ -67,38 +86,131 @@ export default Component.extend({
...
+  /**
+   * converts controls to array if they are given as a hash
+   * filters out inapplicable controls
+   * @param {object|array} controls - array or hash of controls
+   * @param {array} selectedItems - array of selected items for filtering
+   * @returns {array} - array of controls
+   */
+  getNormalizedControls (controls, selectedItems) {
+    // convert hash of controls to array if necessary
+    // notice this is Ember typeOf, which is very differnt from native typeof
```
Comment (notmessenger) [inline addon/components/frost-action-bar.js]: `differnt` -> `different`

Though don't think we need this comment.

File: addon/components/frost-action-bar.js (hunk 1 above)
Comment (gkchestertron) [reply addon/components/frost-action-bar.js]: ty

File: addon/components/frost-action-bar.js (hunk 3 above)
Comment (gkchestertron) [reply addon/components/frost-action-bar.js]: I accepts either frost-buttons or plain objects. This is why I used PropTypes.object because specifying both here only really serves as documentation.

File: tests/integration/components/frost-action-bar-test.js (hunk 9 above)
Comment (gkchestertron) [reply tests/integration/components/frost-action-bar-test.js]: ty

File: README.md (hunk 2 above)
Comment (gkchestertron) [reply README.md]: yeah - the naming is funny. I was following what happened in forch. But, perhaps it would make more sense to call the property just 'moreActions'.

File: addon/components/frost-action-bar.js (hunk 10 above)
Comment (gkchestertron) [reply addon/components/frost-action-bar.js]: This is something I wanted to ask you about, but you were gone by time I got to it on Friday. So, this does work, I believe because of the way getDefaultProps works--but not really a great approach. I didn't want to use a different prop because it seemed strange to have two props for its visibility. But, isVisible isn't part of the interface... which for some reason wasn't occurring to me at the time. I'm going to blame this on it being Friday at the time. I think I'll just move it to "forceVisibility" and use that in the isVisible computed prop.

File: addon/components/frost-action-bar.js (hunk 12 above)
Comment (gkchestertron) [reply addon/components/frost-action-bar.js]: prob not, but it tripped me up first time I came to it.

File: addon/components/frost-action-bar.js (hunk 10 above)
Comment (gkchestertron) [reply addon/components/frost-action-bar.js]: And add a test for it.

File: addon/components/frost-action-bar.js (hunk 1 above)
Comment (gkchestertron) [reply addon/components/frost-action-bar.js]: Also, fixed the spellchecker for my comments...

File: README.md (hunk 2 above)
Comment (lirwin) [reply README.md]: yes i prefer `moreActions` and `moreActionsText`

File: tests/unit/components/frost-action-bar-test.js [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        })
+      })
+    })
+
+    describe('isFrostButton', function () {
+      it('should return true if component is a frost button', function () {
+        expect(component.isFrostButton(button212)).to.equal(true)
+      })
+
+      it('should return false if component is a frost button', function () {
```
Comment (notmessenger) [inline tests/unit/components/frost-action-bar-test.js]: is _not_ a frost button?

File: tests/unit/components/frost-action-bar-test.js [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        })
+      })
+    })
+
+    describe('isFrostButton', function () {
+      it('should return true if component is a frost button', function () {
+        expect(component.isFrostButton(button28)).to.equal(true)
+      })
+
+      it('should return false if component is a frost button', function () {
```
Comment (notmessenger) [inline tests/unit/components/frost-action-bar-test.js]: is _not_ a frost button?

File: README.md [hunk 15]
```diff
@@ -26,13 +26,16 @@ ember install ember-frost-object-browser
...
+| Attribute       | Type                                        | Value | Description                                         |
+| --------------- | ----------------------------------          | ----- | --------------------------------------------------- |
+| controls        | `array`                                     |       | Controls that will be available in the action bar   |
+| isLoading       | `boolean`                                   | false | **default** - Action bar is not in a loading state  |
+|                 |                                             | true  | Action bar is in a loading state                    |
+| forceVisible    | `boolean`                                   |       | forces the component's visibility                   |
+| loadingText     | `string`                                    |       | Text that appears beside the loading animation      |
+| moreActions     | `Ember.Component[]|boolean|object[]|object` |       | Only accepts `frost-button`s if passing components  |
+| moreActiosnText | `string`                                    |       | Text for 'More...' button                           |
```
Comment (notmessenger) [inline README.md]: `moreActiosnText` -> `moreActionsText`

File: README.md [hunk 16]
```diff
@@ -26,13 +26,16 @@ ember install ember-frost-object-browser
...
-| isLoading        | `boolean`     | false | **default** - Action bar is not in a loading state  |
-|                  |               | true  | Action bar is in a loading state                    |
-| loadingText      | `string`      |       | Text that appears beside the loading animation      |
-| selectedItems    | `array`       |       | List of items that are currently selected           |
+| Attribute       | Type                                        | Value | Description                                         |
+| --------------- | ----------------------------------          | ----- | --------------------------------------------------- |
+| controls        | `array`                                     |       | Controls that will be available in the action bar   |
+| isLoading       | `boolean`                                   | false | **default** - Action bar is not in a loading state  |
+|                 |                                             | true  | Action bar is in a loading state                    |
+| forceVisible    | `boolean`                                   |       | forces the component's visibility                   |
```
Comment (notmessenger) [inline README.md]: I don't understand how this fits into this PR.  I didn't see anything mentioned about it in the changelog and it does not seem related to the new behavior.  Is it left over from something else or was it intended to be added as well?  Would `isVisible` align more with `isLoading`?  I'm not sure I'm sold on the `force` part.

**Comment by notmessenger** [comment]
I understand that the action bar is intended to work with `frost-button` components and it is the component we use within the Frost ecosystem but I'm not sure I like the idea of looking for that component type specifically.   By doing so we force the usage to require `frost-button` rather than only needing to be passed a component.   Even if we take such "meta" support off the table for this discussion how does this code hold up if an application has extended the `frost-button` component and is using it instead?

File: addon/components/frost-action-bar.js [hunk 17]
```diff
@@ -80,25 +93,162 @@ export default Component.extend({
...
+          // else grab value from component
+          } else {
+            // props[key] = get(button, `args.named._map.${key}.inner`)
+            props[key] = button.args.named._map[key].inner
+          }
+        }
+
+        return props
+      }, {})
+    // ember 2.8
```
Comment (notmessenger) [inline addon/components/frost-action-bar.js]: https://github.com/ciena-frost/ember-frost-list/blob/master/addon/utils/get-component-name.js has code related to this that may provide additional consideration (I have not evaluated it).

**Comment by notmessenger** [comment]
Please create an issue indicating that the Ember `2.8`-specific introduced in this PR can be removed once no longer supporting Ember `2.8`

File: README.md (hunk 16 above)
Comment (gkchestertron) [reply README.md]: So, by default this component only shows when selected items has length. I looked at a few of our apps, and there are places where the functionality I'm adding will be wanted with the ability to force the action bar component to be visible all the time. I should probably add something about it to the changelog. Perhaps `alwaysVisible` would be better?

File: README.md (hunk 15 above)
Comment (gkchestertron) [reply README.md]: ty

File: tests/unit/components/frost-action-bar-test.js (hunk 13 above)
Comment (gkchestertron) [reply tests/unit/components/frost-action-bar-test.js]: :p

File: tests/unit/components/frost-action-bar-test.js (hunk 14 above)
Comment (gkchestertron) [reply tests/unit/components/frost-action-bar-test.js]: 😛

File: addon/components/frost-action-bar.js (hunk 17 above)
Comment (gkchestertron) [reply addon/components/frost-action-bar.js]: yeah - I'm essentially doing the same thing - perhaps that should be moved to some shared utils in a core repo?

Comment (notmessenger) [reply addon/components/frost-action-bar.js]: Eventually/likely, yes.  It's on my list of things to look at around creating a repo in the frost ecosystem to contain such things for inclusion.

File: README.md (hunk 16 above)
Comment (notmessenger) [reply README.md]: I like `alwaysVisible`.  And do please add an entry in the changelog

**Comment by gkchestertron** [comment]
@notmessenger, your comment about being specific to frost-button is a good point. I was specifically trying to avoid wrapping up links or other dropdowns because this feature is for wrapping up actions. Perhaps I could generalize it to components that have and onClick property.

**Comment by notmessenger** [comment]
That might be an interesting approach to the problem

File: addon/components/frost-action-bar.js (hunk 17 above)
Comment (gkchestertron) [reply addon/components/frost-action-bar.js]: I don't know that we need any more repos... could it be part of frost-core?

Comment (notmessenger) [reply addon/components/frost-action-bar.js]: If every repo we have/use or would need to use this code in already depends on `ember-frost-core` then perhaps but if not then having it separate would be preferred (until tree shaking is complete).

**Comment by notmessenger** [comment]
👍

**Comment by notmessenger** [comment]
👍

File: README.md [hunk 18]
```diff
@@ -136,6 +151,10 @@ onSortingChange (sortOrder) {…}
 
 A scenario where the loading state in the action bar might be useful is if there is some amount of processing that needs to be done in order to determine the state of the controls within the action bar. For example, determining whether a button should be enabled/disabled or shown/hidden. Note that the loading animation is the ring type from [ember-frost-core](https://github.com/ciena-frost/ember-frost-core).
 
+#### More... Button
+
```
Comment (lirwin) [inline README.md]: i was thinking we should document which buttons will get wrapped up into the `More...`.  Doesn't the order in the `controls` array matter?  Does your code exempt the first 4 `frost-button`s it encounters, then take the remaining `frost-button`s and wrap these into the `More...` button?

If so, change documentation to state thus.


take any `frost-button`s and wrap them up into a `More...` button

Comment (notmessenger) [reply README.md]: Good point @lirwin.  @gkchestertron please either make this update or create an issue to capture the need to update the documentation.

File: addon/components/frost-action-bar.js [hunk 19]
```diff
@@ -80,25 +93,162 @@ export default Component.extend({
...
+   * @param {Ember.Object[]|object[]} selectedItems - array of selected items for filtering controls
+   * @param {Ember.Component[]|boolean|object[]|object} moreActions - array or hash of extra controls
+   *  or flag indicating whether to generate them automatically
+   * @returns {Ember.Component[]} array of controls
+   */
+  _controls (controls, selectedItems, moreActions) {
+    controls = this.getNormalizedControls(controls, selectedItems)
+
+    // slice off extra buttons if needed (default)
+    if (moreActions === true) {
```
Comment (lirwin) [inline addon/components/frost-action-bar.js]: sorry comment is not helping me understand this code

isn't `buttonCount` here giving you count of control items which are NOT `frost-button`s?

File: addon/components/frost-action-bar.js [hunk 20]
```diff
@@ -80,25 +93,162 @@ export default Component.extend({
   },
 
   @readOnly
-  @computed('controls')
-  isControlsObject (controls) {
-    return typeOf(controls) === 'object'
+  @computed('controls', 'selectedItems.[]', 'moreActions')
+  /**
+   * gets controls for the action bar
```
Comment (lirwin) [inline addon/components/frost-action-bar.js]: explain here the manipulation being done to the `controls` array with respect to 'button slicing` or preparation for creating the `More...` button if `moreActions` is true

File: README.md (hunk 18 above)
Comment (gkchestertron) [reply README.md]: Good point - issue created to update readme and inline documentation to clarify.
